sdk/python/kfp/compiler/compiler_test.py

```diff
@@ -2652,5 +2631,169 @@ def my_pipeline(x: Input[Artifact] = Artifact(
+            return 2 * num
+
+        @dsl.component
+        def add(nums: List[int]) -> int:
+            return sum(nums)
+
+        @dsl.pipeline
+        def my_pipeline(text: str = ''):
+            with dsl.ParallelFor([1, 2, 3]) as f:
+                with dsl.Condition(text == 'text'):
+                    t = double(num=f)
+
+            with dsl.Condition(text == 'text'):
+                x = add(nums=dsl.Collected(t.output))
+
+    def test_producer_condition_illegal1(self):
+
+        @dsl.component
+        def double(num: int) -> int:
+            return 2 * num
+
+        @dsl.component
+        def add(nums: List[int]) -> int:
+            return sum(nums)
+
+        with self.assertRaisesRegex(
+                compiler_utils.InvalidTopologyException,
+                r'Tasks cannot depend on an upstream task inside Condition that is not a common ancestor of both tasks\. Task add depends on upstream task double\.'
+        ):
+
+            @dsl.pipeline
+            def my_pipeline(text: str = ''):
+                with dsl.Condition(text == 'text'):
+                    with dsl.ParallelFor([1, 2, 3]) as f:
+                        t = double(num=f)
+
+                with dsl.Condition(text == 'text'):
+                    x = add(nums=dsl.Collected(t.output))
+
+    def test_producer_condition_illegal2(self):
+
+        @dsl.component
+        def double(num: int) -> int:
+            return 2 * num
+
+        @dsl.component
+        def add(nums: List[int]) -> int:
+            return sum(nums)
+
+        with self.assertRaisesRegex(
+                compiler_utils.InvalidTopologyException,
+                r'Tasks cannot depend on an upstream task inside Condition that is not a common ancestor of both tasks\. Task add depends on upstream task double\.'
+        ):
+
+            @dsl.pipeline
+            def my_pipeline(text: str = ''):
+                with dsl.Condition(text == 'a'):
+                    with dsl.Condition(text == 'b'):
+                        with dsl.ParallelFor([1, 2, 3]) as f:
+                            t = double(num=f)
+                    add(nums=dsl.Collected(t.output))
+
+    def test_producer_exit_handler_illegal1(self):
+
+        @dsl.component
+        def double(num: int) -> int:
+            return 2 * num
+
+        @dsl.component
+        def exit_comp():
+            print('Running exit task!')
+
+        @dsl.component
+        def add(nums: List[int]) -> int:
+            return sum(nums)
+
+        with self.assertRaisesRegex(
+                compiler_utils.InvalidTopologyException,
+                'Tasks cannot depend on an upstream task inside ExitHandler that is not a common ancestor of both tasks\. Task add depends on upstream task double\.'
+        ):
+
+            @dsl.pipeline
+            def my_pipeline(text: str = ''):
+                with dsl.ExitHandler(exit_comp()):
+                    with dsl.ParallelFor([1, 2, 3]) as f:
```

Comment:
Since you listed out combinations of exithandler-parrallelfor and condition-parallelfor, I'm curious what would be the case  for parrallelfor-parallelfor, it would be interested to see what are valid cases vs. invalid cases under nested loops.

Reply:
I've added test cases for this for completeness in the `compiler_test.py` file.

The actual YAML for these cases is tested extensively in the `read_write_test.py` tests.